src/v/iceberg/schema_protobuf.cc

```diff
@@ -0,0 +1,157 @@
+ */
+using field_outcome = conversion_outcome<nested_field_ptr>;
+using struct_outcome = conversion_outcome<struct_type>;
+
+field_outcome from_protobuf(const pb::FieldDescriptor& fd, bool is_repeated);
+
+field_outcome success(const pb::FieldDescriptor& fd, field_type ft) {
+    return nested_field::create(
+      fd.index(), fd.name(), field_required::no, std::move(ft));
+}
+
+struct_outcome struct_from_protobuf(const pb::Descriptor& msg) {
+    struct_type struct_t;
+    struct_t.fields.reserve(msg.field_count());
+    for (int i = 0; i < msg.field_count(); ++i) {
+        auto field = msg.field(i);
+        auto res = from_protobuf(*field, field->is_repeated());
+        if (res.has_error()) {
+            return res.error();
+        }
+        struct_t.fields.push_back(std::move(res).value());
+    }
+    return std::move(struct_t);
+}
+
+field_outcome from_protobuf(const pb::FieldDescriptor& fd, bool is_repeated) {
+    if (fd.is_map()) {
+        auto mt = fd.message_type();
+        auto key_field = mt->map_key();
+        auto value_field = mt->map_value();
+        auto key_nested_res = from_protobuf(
+          *key_field, key_field->is_repeated());
+        if (key_nested_res.has_error()) {
+            return key_nested_res;
+        }
+        auto value_field_nested_res = from_protobuf(
+          *value_field, value_field->is_repeated());
+        if (value_field_nested_res.has_error()) {
+            return value_field_nested_res;
+        }
+        auto field_type = map_type::create(
+          key_field->index(),
+          std::move(std::move(key_nested_res).assume_value()->type),
+          value_field->index(),
+          field_required(value_field->is_optional()),
+          std::move(std::move(value_field_nested_res).assume_value()->type));
+        return success(fd, std::move(field_type));
+    }
+    if (is_repeated) {
+        auto field_type_res = from_protobuf(fd, false);
+        if (field_type_res.has_error()) {
+            return field_type_res;
+        }
+
+        auto type = list_type::create(
+          fd.index(),
+          field_required::no,
+          std::move(std::move(field_type_res).assume_value()->type));
+        return success(fd, std::move(type));
+    }
+
+    switch (fd.type()) {
+    case pb::FieldDescriptor::Type::TYPE_BOOL:
+        return success(fd, boolean_type{});
+    case pb::FieldDescriptor::TYPE_DOUBLE:
+        return success(fd, double_type{});
+    case pb::FieldDescriptor::TYPE_FLOAT:
+        return success(fd, float_type{});
+    case pb::FieldDescriptor::TYPE_INT64:
+        return success(fd, long_type{});
+    case pb::FieldDescriptor::TYPE_UINT32:
+    case pb::FieldDescriptor::TYPE_FIXED32:
+    case pb::FieldDescriptor::TYPE_UINT64:
+    case pb::FieldDescriptor::TYPE_FIXED64:
+        return schema_conversion_exception(fmt::format(
+          "Protocol buffer field {} type {} not supported, iceberg only "
+          "supports signed integers",
```

Comment:
Groups are good to not handle because they have been deprecated for more than a decade. But these are full supported and customers may be locked into these values for various reasons. Can we consider supporting them? Let me do research and see what other products do here

Reply:
So java does not have unsigned type, and all this ecosystem is java centric, so it's probably fine to adopt java semantics for protocol buffers and just cast to a signed int - this approach doesn't lose bits so it's generally acceptable.

Reply:
Here is what official iceberg does for parquet: https://github.com/apache/iceberg/blob/79fd977f67592a16579cff31478e7ea98ef126e4/parquet/src/main/java/org/apache/iceberg/parquet/MessageTypeToType.java#L223

Honestly I don't understand the bit about not supporting uint64 but they convert uint32 into iceberg Long type.

Reply:
i guess the `int64` is still able to hold the `uint32_t` max value wheres casting `uint64` to `int64_t` may result in overflow. I will implement the same semantics.